packages/async-flow/src/endowments.js

```diff
@@ -0,0 +1,261 @@
+      }
+      if (Array.isArray(e)) {
+        return 'array';
+      }
+      const keys = ownKeys(e);
+      keys.length >= 1 || Fail`empty record should be storable ${e}`;
+      const desc = getOwnPropertyDescriptor(e, keys[0]);
+      assert(desc !== undefined); // just for type checker
+      if ('value' in desc) {
+        return 'record';
+      } else {
+        'get' in desc || Fail`internal: unexpected descriptor ${desc}`;
+        return 'state';
+      }
+    } else {
+      throw Fail`unexpected endowment ${e}`;
+    }
+  };
+  harden(endowmentKindOf);
+
+  const unwrapMap = outerZone.weakMapStore('unwrapMap', {
+    keyShape: M.remotable('wrapped'),
+    valueShape: M.remotable('unwrapper'),
+  });
+
+  const unwrapMapHas = k => {
+    if (isVow(k) || isRemotable(k)) {
+      return unwrapMap.has(toPassableCap(k));
+    } else {
+      return false;
+    }
+  };
+  const unwrapMapGet = k => unwrapMap.get(toPassableCap(k));
+  const unwrapMapSet = (k, v) => {
+    const k2 = toPassableCap(k);
+    if (unwrapMapHas(k)) {
+      unwrapMap.set(k2, v);
+    } else {
+      unwrapMap.init(k2, v);
+    }
+  };
+
+  /**
+   * @param {Zone} zone
+   * @param {string} tag
+   * @param {unknown} e
+   */
+  const prepareEndowment = (zone, tag, e) => {
+    const eKind = endowmentKindOf(e);
+    switch (eKind) {
+      case 'promise': {
+        const p = /** @type {Promise} */ (e);
+        // Not using watch or watchPromise because upgrade rejection
+        // should leave it to be resolved by the next promise endowment
+        const { vow, resolver } = makeVowKit();
+        void E.when(
+          p,
+          v => {
+            resolver.resolve(v);
+          },
+          reason => {
+            if (!isUpgradeDisconnection(reason)) {
+              resolver.reject(reason);
+            }
+          },
+        );
+        return vow;
+      }
+      case 'storable': {
+        return e;
+      }
+      case 'far': {
+        const r = /** @type {RemotableObject} */ (e);
+        const methods = forwardingMethods(r);
+        return zone.exo(
+          tag,
+          M.interface('FarWrapped', {}, { defaultGuards: 'raw' }),
+          methods,
+        );
+      }
+      case 'function': {
+        const f = /** @type {Callable} */ (e);
+        const wrapped = zone.exo(tag, FunctionWrapperI, {
+          apply(args) {
+            return f(...args);
+          },
+        });
+        unwrapMapSet(wrapped, functionUnwrapper);
+        return wrapped;
+      }
+      case 'array': {
+        const a = /** @type {unknown[]} */ (e);
+        const subZone = zone.subZone(tag);
+        return a.map((subE, i) => prepareEndowment(subZone, `${i}`, subE));
+      }
+      case 'record': {
+        const r = /** @type {Record<PropertyKey, unknown>} */ (e);
+        const subZone = zone.subZone(tag);
+        return objectMap(r, (subE, k) =>
+          prepareEndowment(subZone, String(k), subE),
+        );
+      }
+      case 'state': {
```

Comment:
Does this implement Exo instance state?!

Reply:
Not quite, but it lays the groundwork. That's why I wrote it that way. I'm already writing that next step in my head. I don't yet know the degree to which we can realistically say we support asyncFlow async methods, but we'll see. Don't expect that one soon.

Resolving.